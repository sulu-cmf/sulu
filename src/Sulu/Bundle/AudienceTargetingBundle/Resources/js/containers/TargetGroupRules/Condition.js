// @flow
import React from 'react';
import {Button, SingleSelect} from 'sulu-admin-bundle/components';
import ruleRegistry from './registries/ruleRegistry';
import ruleTypeRegistry from './registries/ruleTypeRegistry';
import conditionStyles from './condition.scss';
import type {Condition as ConditionType} from './types';

type Props = {|
    index: number,
    onChange: (value: ConditionType, index: number) => void,
    onRemove: (index: number) => void,
    value: ConditionType,
|};

class Condition extends React.Component<Props> {
    handleRuleTypeChange = (rule: string) => {
        const {index, onChange, value} = this.props;
        onChange({...value, type: rule}, index);
    };

    handleRuleChange = (condition: ConditionType) => {
        const {index, onChange, value} = this.props;
        onChange({...value, condition}, index);
    };

    handleRemove = () => {
        const {index, onRemove} = this.props;
        onRemove(index);
    };

    render() {
        const {value} = this.props;

        const type = value.type ? ruleRegistry.get(value.type).type : undefined;
        const RuleType = type ? ruleTypeRegistry.get(type.name) : undefined;

        return (
            <div className={conditionStyles.conditionContainer}>
                <div className={conditionStyles.condition}>
                    <div className={conditionStyles.select}>
                        <SingleSelect onChange={this.handleRuleTypeChange} value={value.type}>
                            {Object.keys(ruleRegistry.getAll()).map((ruleKey) => (
                                <SingleSelect.Option key={ruleKey} value={ruleKey}>
                                    {ruleRegistry.get(ruleKey).name}
                                </SingleSelect.Option>
                            ))}
                        </SingleSelect>
                    </div>
                    <div className={conditionStyles.type}>
                        {!!RuleType &&
                            <RuleType
                                onChange={this.handleRuleChange}
                                options={type && type.options}
                                value={value.condition}
                            />
                        }
                    </div>
                </div>
                <Button className={conditionStyles.icon} icon="su-trash-alt" onClick={this.handleRemove} skin="icon" />
            </div>
        );
    }
}

export default Condition;
