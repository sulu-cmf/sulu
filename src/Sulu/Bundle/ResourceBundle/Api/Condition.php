<?php
/*
 * This file is part of the Sulu CMS.
 *
 * (c) MASSIVE ART WebServices GmbH
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Sulu\Bundle\ResourceBundle\Api;

use Sulu\Bundle\ResourceBundle\Resource\DataTypes;
use Sulu\Component\Rest\ApiWrapper;
use Sulu\Bundle\ResourceBundle\Entity\Condition as ConditionEntity;
use JMS\Serializer\Annotation\VirtualProperty;
use JMS\Serializer\Annotation\SerializedName;
use JMS\Serializer\Annotation\ExclusionPolicy;

/**
 * The Condition class which will be exported to the API
 *
 * @package Sulu\Bundle\ResourceBundle\Api
 * @ExclusionPolicy("all")
 */
class Condition extends ApiWrapper
{
    /**
     * @param ConditionEntity $entity
     * @param string $locale
     */
    public function __construct(ConditionEntity $entity, $locale)
    {
        $this->entity = $entity;
        $this->locale = $locale;
    }

    /**
     * Get id
     *
     * @VirtualProperty
     * @SerializedName("id")
     * @return integer
     */
    public function getId()
    {
        return $this->entity->getId();
    }

    /**
     * Get value
     *
     * @VirtualProperty
     * @SerializedName("value")
     * @return mixed
     */
    public function getValue()
    {
        // return a proper date if it is a datetime value
        if($this->getType() === DataTypes::DATETIME_TYPE) {
            return new \DateTime($this->entity->getValue());
        }
        return $this->entity->getValue();
    }

    /**
     * Get field
     *
     * @VirtualProperty
     * @SerializedName("field")
     * @return string
     */
    public function getField()
    {
        return $this->entity->getField();
    }

    /**
     * Get operator
     *
     * @VirtualProperty
     * @SerializedName("operator")
     * @return string
     */
    public function getOperator()
    {
        return $this->entity->getOperator();
    }

    /**
     * Get type
     *
     * @VirtualProperty
     * @SerializedName("type")
     * @return integer
     */
    public function getType()
    {
        return $this->entity->getType();
    }

    /**
     * Set field
     *
     * @param string $field
     */
    public function setField($field)
    {
        $this->entity->setField($field);
    }

    /**
     * Set operator
     *
     * @param string $operator
     */
    public function setOperator($operator)
    {
        $this->entity->setOperator($operator);
    }

    /**
     * Set type
     *
     * @param integer $type
     */
    public function setType($type)
    {
        $this->entity->setType($type);
    }

    /**
     * Set value
     *
     * @param string $value
     */
    public function setValue($value)
    {
        $this->entity->setValue($value);
    }

    /**
     * Set conditionGroup
     *
     * @param ConditionGroup $conditionGroup
     */
    public function setConditionGroup(ConditionGroup $conditionGroup)
    {
        $this->entity->setConditionGroup($conditionGroup->getEntity());
    }

    /**
     * Get conditionGroup
     *
     * @return ConditionGroup
     */
    public function getConditionGroup()
    {
        return new ConditionGroup($this->entity->getConditionGroup(), $this->locale);
    }
}
