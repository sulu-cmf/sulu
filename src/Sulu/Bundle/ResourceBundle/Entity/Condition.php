<?php

namespace Sulu\Bundle\ResourceBundle\Entity;

use Doctrine\ORM\Mapping as ORM;

/**
 * Condition
 */
class Condition
{
    const TYPE_STRING = 1;
    const TYPE_NUMBER = 2;
    const TYPE_DATETIME = 3;

    /**
     * @var string
     */
    private $field;

    /**
     * @var string
     */
    private $operator;

    /**
     * @var integer
     */
    private $type;

    /**
     * @var string
     */
    private $value;

    /**
     * @var integer
     */
    private $id;

    /**
     * @var \Sulu\Bundle\ResourceBundle\Entity\ConditionGroup
     */
    private $conditionGroup;


    /**
     * Set field
     *
     * @param string $field
     * @return Condition
     */
    public function setField($field)
    {
        $this->field = $field;

        return $this;
    }

    /**
     * Get field
     *
     * @return string 
     */
    public function getField()
    {
        return $this->field;
    }

    /**
     * Set operator
     *
     * @param string $operator
     * @return Condition
     */
    public function setOperator($operator)
    {
        $this->operator = $operator;

        return $this;
    }

    /**
     * Get operator
     *
     * @return string 
     */
    public function getOperator()
    {
        return $this->operator;
    }

    /**
     * Set type
     *
     * @param integer $type
     * @return Condition
     */
    public function setType($type)
    {
        $this->type = $type;

        return $this;
    }

    /**
     * Get type
     *
     * @return integer 
     */
    public function getType()
    {
        return $this->type;
    }

    /**
     * Set value
     *
     * @param string $value
     * @return Condition
     */
    public function setValue($value)
    {
        $this->value = $value;

        return $this;
    }

    /**
     * Get value
     *
     * @return string 
     */
    public function getValue()
    {
        return $this->value;
    }

    /**
     * Get id
     *
     * @return integer 
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * Set conditionGroup
     *
     * @param \Sulu\Bundle\ResourceBundle\Entity\ConditionGroup $conditionGroup
     * @return Condition
     */
    public function setConditionGroup(\Sulu\Bundle\ResourceBundle\Entity\ConditionGroup $conditionGroup)
    {
        $this->conditionGroup = $conditionGroup;

        return $this;
    }

    /**
     * Get conditionGroup
     *
     * @return \Sulu\Bundle\ResourceBundle\Entity\ConditionGroup 
     */
    public function getConditionGroup()
    {
        return $this->conditionGroup;
    }
}
