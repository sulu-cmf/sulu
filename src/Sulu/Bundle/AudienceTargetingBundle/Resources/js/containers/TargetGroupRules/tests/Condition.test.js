// @flow
import React from 'react';
import {render, shallow} from 'enzyme';
import ruleRegistry from '../registries/ruleRegistry';
import ruleTypeRegistry from '../registries/ruleTypeRegistry';
import Condition from '../Condition';

jest.mock('sulu-admin-bundle/utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

jest.mock('../registries/ruleRegistry', () => {
    const getAllMock = jest.fn();

    return {
        getAll: getAllMock,
        get: (key) => getAllMock()[key],
    };
});

jest.mock('../registries/ruleTypeRegistry', () => ({
    get: jest.fn(),
}));

test('Render a condition', () => {
    ruleRegistry.getAll.mockReturnValue({
        browser: {
            name: 'Browser',
            type: {
                name: 'select',
                options: {},
            },
        },
    });

    const value = {condition: {}, type: 'browser'};
    expect(render(<Condition index={1} onChange={jest.fn()} onRemove={jest.fn()} value={value} />)).toMatchSnapshot();
});

test('Call onRemove callback if remove icon is clicked', () => {
    const removeSpy = jest.fn();
    const value = {condition: {}, type: undefined};

    const condition = shallow(<Condition index={5} onChange={jest.fn()} onRemove={removeSpy} value={value} />);
    condition.find('Button[icon="su-trash-alt"]').prop('onClick')();

    expect(removeSpy).toBeCalledWith(5);
});

test('Call onChange callback if rule type changes', () => {
    const changeSpy = jest.fn();
    const value = {condition: {test: 'value'}, type: undefined};

    const condition = shallow(<Condition index={5} onChange={changeSpy} onRemove={jest.fn()} value={value} />);
    condition.find('SingleSelect').prop('onChange')('browser');

    expect(changeSpy).toBeCalledWith({condition: {test: 'value'}, type: 'browser'}, 5);
});

test('Call onChange callback if condition changes', () => {
    ruleRegistry.getAll.mockReturnValue({
        browser: {
            name: 'Browser',
            type: {
                name: 'select',
                options: {},
            },
        },
    });

    const RuleType = () => null;
    ruleTypeRegistry.get.mockImplementation((key) => {
        switch (key) {
            case 'select':
                return RuleType;
        }
    });

    const changeSpy = jest.fn();
    const value = {condition: {}, type: 'browser'};

    const condition = shallow(<Condition index={5} onChange={changeSpy} onRemove={jest.fn()} value={value} />);
    condition.find(RuleType).prop('onChange')({test: 'value'});

    expect(changeSpy).toBeCalledWith({condition: {test: 'value'}, type: 'browser'}, 5);
});
