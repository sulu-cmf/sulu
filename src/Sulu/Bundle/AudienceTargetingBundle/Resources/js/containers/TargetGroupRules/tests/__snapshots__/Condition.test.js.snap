// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a condition 1`] = `
<div
  class="conditionContainer"
>
  <div
    class="condition"
  >
    <div
      class="select"
    >
      <div
        class="select"
      >
        <button
          class="displayValue default"
          type="button"
        >
          <div
            aria-label="Browser"
            class="croppedText"
            title="Browser"
          >
            <div
              aria-hidden="true"
              class="front"
            >
              Brow
            </div>
            <div
              aria-hidden="true"
              class="back"
            >
              <span>
                ser
              </span>
            </div>
            <div
              class="whole"
            >
              Browser
            </div>
          </div>
          <span
            aria-label="su-angle-down"
            class="su-angle-down toggle"
          />
        </button>
      </div>
    </div>
    <div
      class="type"
    />
  </div>
  <button
    class="button icon icon"
    type="button"
  >
    <span
      aria-label="su-trash-alt"
      class="su-trash-alt buttonIcon"
    />
  </button>
</div>
`;
